crates/fluvio-hub-util/src/fvm/api/download.rs

```diff
@@ -0,0 +1,186 @@
+//! Download API for downloading the artifacts from the server
+
+use std::path::{PathBuf};
+use std::io::{Cursor, copy, Read};
+use std::fs::File;
+
+use anyhow::{Error, Result};
+use http_client::async_trait;
+use sha2::{Sha256, Digest};
+use surf::StatusCode;
+
+use crate::fvm::Artifact;
+
+/// Generates the Sha256 checksum for the specified file
+fn shasum256(file: &File) -> Result<String> {
+    let meta = file.metadata()?;
+    let mut hasher = Sha256::new();
+    let mut buffer = vec![0; meta.len() as usize];
+
+    file.read_exact(&mut buffer)?;
+    hasher.update(buffer);
+
+    let output = hasher.finalize();
+    Ok(hex::encode(output))
+}
+
+/// Verifies downloaded artifact checksums against the upstream checksums
+async fn checksum(artf: &Artifact, path: &PathBuf) -> Result<()> {
+    let file = File::open(path)?;
+    let local_file_shasum = shasum256(&file)?;
+    let upstream_shasum = surf::get(&artf.sha256_url)
+        .await
+        .map_err(|err| Error::msg(err.to_string()))?
+        .body_string()
+        .await
+        .map_err(|err| Error::msg(err.to_string()))?;
+
+    if local_file_shasum != upstream_shasum {
+        return Err(Error::msg(format!(
+            "Artifact {} didnt matched upstream shasum. {} != {}",
+            artf.name, local_file_shasum, upstream_shasum
+        )));
+    }
+
+    Ok(())
+}
+
+#[async_trait]
+pub trait Download {
+    /// Downloads the artifact to the specified directory
+    ///
+    /// Internally validates the checksum of the downloaded artifact
+    async fn download(&self, target_dir: PathBuf) -> Result<()>;
+}
+
+#[async_trait]
+impl Download for Artifact {
+    async fn download(&self, target_dir: PathBuf) -> Result<()> {
+        let mut res = surf::get(&self.download_url)
+            .await
+            .map_err(|err| Error::msg(err.to_string()))?;
+
+        if res.status() == StatusCode::Ok {
+            let out_path = target_dir.join(&self.name);
+            let mut file = File::create(&out_path)?;
+            let mut buf = Cursor::new(
+                res.body_bytes()
+                    .await
+                    .map_err(|err| Error::msg(err.to_string()))?,
+            );
+
+            copy(&mut buf, &mut file)?;
+            checksum(self, &out_path).await?;
+
+            tracing::debug!(
+                "Artifact downloaded: {} at {:?}",
+                self.name,
+                out_path.display()
+            );
+
+            return Ok(());
+        }
+
+        Err(Error::msg(format!(
+            "Server responded with Status Code {}",
+            res.status()
+        )))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use tempfile::TempDir;
+
+    use super::*;
+
+    #[async_std::test]
```

Comment:
use `fluvio::test`
also since this is integration test, mark with ignore

Reply:
`#[fluvio_future::test]`! Sure thing!